apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java

```diff
@@ -0,0 +1,152 @@
+package com.ctrip.framework.apollo.core.utils;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.slf4j.Logger;
+import org.slf4j.event.Level;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Delayed log printing utility class, used only for logging when Apollo is initialized
+ *
+ * @author kl (http://kailing.pub)
+ * @since 2021/5/11
+ */
+final class DeferredLogCache {
+
+  public static final int MAX_LOG_SIZE = 1000;
+  private static final AtomicInteger LOG_INDEX = new AtomicInteger(0);
+  private static final Cache<Integer, Line> LOG_CACHE = CacheBuilder.newBuilder()
+      .maximumSize(MAX_LOG_SIZE)
+      .expireAfterWrite(1, TimeUnit.MINUTES)
+      .build();
+
+  private DeferredLogCache() {
+  }
+
+
+  public static void debug(Logger logger, String message, Throwable throwable, Object... objects) {
+    add(logger, Level.DEBUG, message, objects, throwable);
+  }
+
+  public static void info(Logger logger, String message, Throwable throwable, Object... objects) {
+    add(logger, Level.INFO, message, objects, throwable);
+  }
+
+  public static void warn(Logger logger, String message, Throwable throwable, Object... objects) {
```

Comment:
with this signature, we can't make sure throwable and objects are not null at the same time, maybe we could add another overload method?

```java
public static void warn(Logger logger, String message, Object... objects) {}
public static void warn(Logger logger, String message, Throwable throwable) {}
```

Reply:
btw, looks like the ci failed, please take a look~

Reply:
DeferredLogCache is not open to the public, the current api meets the needs, the compilation error is caused by my negligence

Reply:
OK, then I would suggest adding a method javadoc here to make it clear that user should not pass throwable and objects at the same time in case it was ignored in the future